.github/workflows/dependencies.yaml

```diff
@@ -16,20 +21,29 @@ jobs:
       image: ghcr.io/matter-labs/zksync-llvm-runner:latest
 
     steps:
+      # Ref to checkout will be defined automatically from the event that triggered the workflow
       - name: Checkout source
         uses: actions/checkout@v4
-        with:
-          ref: ${{ inputs.dependencies_branch }}
-
-      - name: Patch era-compiler-common
-        run: |
-          echo "[patch.\"https://github.com/matter-labs/era-compiler-common\"]" >> "Cargo.toml"
-          echo "era-compiler-common = { git = 'https://www.github.com/matter-labs/era-compiler-common', branch = '${{ inputs.dependencies_branch }}' }" >> "Cargo.toml"
 
-      - name: Patch era-compiler-llvm-context
+      - name: Patch dependencies
+        shell: bash -ex {0}
+        env:
+          CARGO_TOML: Cargo.toml
+          GH_ML_URL: https://github.com
+          # `www.github.com` is a workaround for Cargo issue with `patch` section from `git` source
+          # https://github.com/rust-lang/cargo/issues/5478
+          GH_ML_PATCHED_URL: https://www.github.com
         run: |
-          echo "[patch.\"https://github.com/matter-labs/era-compiler-llvm-context\"]" >> "Cargo.toml"
-          echo "era-compiler-common = { git = 'https://www.github.com/matter-labs/era-compiler-llvm-context', branch = '${{ inputs.dependencies_branch }}' }" >> "Cargo.toml"
+          for DEP in ${{ inputs.dependencies_list }}; do
```

Comment:
Hmm, looks really magical.
Could you explain a bit what is happening here?

Reply:
Let me explain a bit what's going on here.

1. This workflow would be triggered from our dependency repositories, e.g. from `era-compiler-common` or `era-compiler-llvm-context`. For each dependency to be _transitively propagated_ in `cargo`, we need to add the following section:

```
[patch.\"https://github.com/matter-labs/era-compiler-common\"]
era-compiler-common = { git = 'https://www.github.com/matter-labs/era-compiler-common', branch = '${{ inputs.dependencies_branch }}' }
```

It means "instead of the default `main`, please use some `${{ inputs.dependencies_branch }}` for this build."

As we can have changes in the multiple dependencies simultaneously, we do this in the `for` loop to update each one of them.

I tried to give a detailed explanation in the [Notion guide](https://www.notion.so/matterlabs/Dependency-testing-bef944f621d642659816c53b541a2281?pvs=4). Please, take a look. I believe you would have some questions / improvements that I can address.